src/pycsou/opt/solver/cg.py

```diff
@@ -0,0 +1,142 @@
+        r"""
+        Solve the minimization problem defined in :py:meth:`CG.__init__`, with the provided
+        run-specific parameters.
+        Parameters
+        ----------
+        x0: NDArray
+           (..., N) Starting guess(es) for the solution(s). Defaults to a zero NDArray if unspecified.
+        mode: Mode
+           Execution mode. See :py:class:`Solver` for usage examples.
+           Useful method pairs depending on the execution mode:
+           * BLOCK: fit()
+           * ASYNC: fit() + busy() + stop()
+           * MANUAL: fit() + steps()
+        tol: Real
+           Tolerance for convergence, norm(residual) <= tol.  Defaults to :math:`1e-5` if unspecified.
+        """
+        # Create a stopping criteria from opt, using tol, and then call super().fit() method
+
+        xp = pycu.get_array_module(self._b)
+        x0 = x0 if (x0 is None) else pycrt.coerce(x0)
+        try:
+            assert not ((stop_crit is None) and (tol is None))
+        except:
+            raise ValueError(f"Both `stop_crit` and `tol` cannot be left undefined. Please define one.")
+        if stop_crit is None:
+            try:
+                assert tol > 0
+            except:
+                raise ValueError(f"tol must be positive, got {tol}.")
+            stop_crit = pycos.AbsError(eps=tol, var="r")
+
+        elif (stop_crit is not None) and (tol is not None):
+            warnings.warn(
+                "The `tol` tolerance value given will be ignored. A custom-made Stopping Criterion has been "
+                "given as argument. To avoid this warning use `tol=None` or use the default stopping "
+                "criterion"
+            )
+        if x0 is not None:
+            try:
+                assert is_broadcastable(x0, self._b)
+
+            except:
+                raise ValueError(
+                    f"Input initial guess has a mismatch in its shape dimension with data array `b` "
+                    f"(shape {x0.shape} is not broadcastable with {self._b.shape})."
+                )
+            try:
+                assert pycu.get_array_module(x0) == xp
+            except:
+                raise ValueError(
+                    f"Input initial guess has a mismatch in its shape array module with data array `b` "
+                    f"(array module {pycu.get_array_module(x0)} is different from {xp}."
+                )
+        else:
+            x0 = xp.zeros((1, self._b.shape[-1]), dtype=pycrt.getPrecision().value)
+
+        return super().fit(x0, mode=mode, stop_crit=stop_crit)
+
+    def m_init(self, x0: pyct.NDArray, tol: float = 1e-5):
+        self._mstate["x"] = x0
+        self._mstate["r"] = r = self._b - self._a.apply(x0)
+        self._mstate["p"] = r.copy()
+
+    def m_step(self):
+        mst = self._mstate
+        x = mst["x"]
+        r = mst["r"]
+        p = mst["p"]
+
+        ap = self._a.apply(p)
+        rr = (r * r).sum(-1)[..., None]
+        alpha = rr / (p * ap).sum(-1)[..., None]
```

Comment:
`rr / (p*ap).sum(axis=-1, keepdims=True)` avoids having to do the `[...,None]` indexing at the end.